File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/io/HttpConfigurable.java
package com.sedmelluq.discord.lavaplayer.tools.io;

import org.apache.http.client.config.RequestConfig;
import org.apache.http.impl.client.HttpClientBuilder;

import java.util.function.Consumer;
import java.util.function.Function;

/**
 * Represents a class where HTTP request configuration can be changed.
 */
public interface HttpConfigurable {
  /**
   * @param configurator Function to reconfigure request config.
   */
  void configureRequests(Function<RequestConfig, RequestConfig> configurator);

  /**
   * @param configurator Function to reconfigure HTTP builder.
   */
  void configureBuilder(Consumer<HttpClientBuilder> configurator);
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/io/HttpConfigurable.kt
package com.sedmelluq.discord.lavaplayer.tools.io

import org.apache.http.client.config.RequestConfig
import org.apache.http.impl.client.HttpClientBuilder

/**
 * Function to reconfigure a [RequestConfig].
 */
typealias RequestConfigurator = RequestConfig.() -> RequestConfig

/**
 * Function to reconfigure an [HttpClientBuilder].
 */
typealias BuilderConfigurator = HttpClientBuilder.() -> Unit

/**
 * Represents a class where HTTP request configuration can be changed.
 */
interface HttpConfigurable {
    /**
     * @param configurator Function to reconfigure request config.
     */
    fun configureRequests(configurator: RequestConfigurator)

    /**
     * @param configurator Function to reconfigure HTTP builder.
     */
    fun configureBuilder(configurator: BuilderConfigurator)
}
